python-package/lightgbm/sklearn.py

```diff
@@ -812,6 +848,15 @@ def predict(self, X, raw_score=False, start_iteration=0, num_iteration=None,
         ):
             predict_params.pop(alias, None)
         predict_params.update(kwargs)
+
+        # number of threads can have values with special meaning which is only applied
+        # in the scikit-learn interface, these should not reach the c++ side as-is
+        n_jobs = self.n_jobs
+        for alias in _ConfigAliases.get("num_threads"):
+            if alias in predict_params:
+                n_jobs = predict_params.pop(alias)
+        predict_params["num_threads"] = self._process_n_jobs(n_jobs)
```

Comment:
This can be simplified with `_choose_param_value()`:
```suggestion
        predict_params = _choose_param_value("num_threads", predict_params, self.n_jobs)
        predict_params["num_threads"] = self._process_n_jobs(predict_params["num_threads"])
```

Reply:
Problem is, `_choose_param_value` treats `None` as if it were not passed, whereas after this PR, `None` has a special value. For example, one can pass `n_jobs=2` in the constructor, then `n_jobs=None` in the call to `predict`, and if that function were used, it would not have the desired effect.

Reply:
Ah, I see now! Thanks for the explanation!

Reply:
@jameslamb Maybe we can refactor `_choose_param_value()` function so that it will handle `None` value of a param correctly?

Reply:
Ok yeah, I think I understand. Will put up a PR shortly.

Reply:
created #5289